storage/replica.go

```diff
@@ -841,6 +851,14 @@ func (r *Replica) applyRaftCommand(ctx context.Context, index uint64, originNode
 	} else {
 		// Update cached appliedIndex if we were able to set the applied index on disk.
 		atomic.StoreUint64(&r.appliedIndex, index)
+		if args.Method() == proto.TruncateLog {
+			ts := proto.RaftTruncatedState{}
+			_, err := engine.MVCCGetProto(r.rm.Engine(), keys.RaftTruncatedStateKey(r.Desc().RangeID),
+				proto.ZeroTimestamp, true, nil, &ts)
+			if err == nil {
```

Comment:
Even in the successful case, this appears to be essentially the same code that `raftTruncatedState()` uses to retrieve the state if it is not cached.  This method should probably just always invalidate the cache with `setTruncatedState(nil)`